### ISSUE TITLE
Webhook gets 403 response

### ISSUE THREAD

**Issue by tastytea** [OpenUserJS/OpenUserJS.org#1610]
I'm trying to sync <https://github.com/tastytea/userscripts> with <https://openuserjs.org/users/tastytea>. The initial ping worked but on every push the error 403 is returned along with these headers:

```
Connection: keep-alive
Content-Length: 2
Date: Wed, 29 May 2019 01:07:36 GMT
Strict-Transport-Security: max-age=31536000000; includeSubDomains
Transfer-Encoding: chunked
Vary: X-HTTP-Method-Override
X-Powered-By: Express
```

![screenshot_2019-05-29T03:20:59](https://user-images.githubusercontent.com/3681516/58522473-d3492180-81c0-11e9-9281-23d9b87d5e31.png)

## Request

```
Request URL: https://openuserjs.org/github/hook
Request method: POST
content-type: application/x-www-form-urlencoded
Expect: 
User-Agent: GitHub-Hookshot/8a8ac33
X-GitHub-Delivery: 83569c8c-81ae-11e9-8822-0daa3e3a04bb
X-GitHub-Event: push
```

[payload.txt](https://github.com/OpenUserJS/OpenUserJS.org/files/3230339/payload.txt)

**Comment by Martii** [comment]
Hmmm... here's my test a few moments ago:

``` console
75547608-81e6-11e9-9ea2-dccfb738f10f … 2019-05-29 08:50:42
Completed in 0.06 seconds.
Request  202 Response
```

https://github.com/Martii/UserScripts/commit/b27841dcb491f1491558364f2fbca285a49ec594 sync'd with https://openuserjs.org/scripts/Marti/RFC_2606%C2%A73_-_Hello,_World!

You appear to be in sync as well on your one script published on OUJS. Your settings for the webhook match mine.

You might consider running a couple more sync tests but please be aware that GH has a caching mechanism in place that will prevent it from serving the "next" version until it has timed out. Usually, from historical brief observation, that is about 10ish minutes *(could be longer or shorter by now)*. So if you do decide to test it be aware we'll reject the update if it's the same content being served. Since it takes time to process the list that GH sends we currently return a 202 as "Accepted" which also closes the connection between GH and OUJS so it's not a resource drain for them and you. About the time of your issue creation the server was running a backup so perhaps the DB was possibly locked too... those are usually fairly brief though. GH could have had a hiccup too... they do that every once in a while with their load balancing.

Anyhow... need you to please retest it with something as it appears to be working well for me.

Thanks.

**Comment by Martii** [comment]
I see you had a deprecation you didn't handle until https://github.com/tastytea/userscripts/commit/d5bef5afa7b4e2c4de416cd4c67ca581f13ea6d5 with:

``` console
@license` is not OSI primary and compatible in the metadata block(s).
```

Perhaps this was your issue?

**Comment by tastytea** [comment]
I tested again at 10:00, 10:39 and 11:20 UTC today. I still get 403 errors back.

> You appear to be in sync as well on your one script published on OUJS

I updated it manually when the automatic sync didn't work.

I thought the issue was maybe that I imported the script manually at first so I created a [new script](https://github.com/tastytea/userscripts/blob/main/test.user.js) for testing, but that didn't change anything.

**Comment by tastytea** [comment]
This is just a shot in the dark, but could it have something to do with my repository not having a master-branch? The default-branch is called “main”.

**Comment by Martii** [comment]
> ... not having a master-branch? The default-branch is called “main”

That's a very good chance... see [this current fixed restriction](https://github.com/OpenUserJS/OpenUserJS.org/blob/a55aa7bad0fb8afcf99796b968f37dbc8f6e5f78/controllers/scriptStorage.js#L2034-L2038). Nice aim btw. :smile_cat: 

I think I'll add some of these into the site docs so it should be clearer esp. when I haven't revisited that section in a while.

**Comment by tastytea** [comment]
The payload has keys named `repository.default_branch` and `repository.master_branch`, both set to "main" in my case. Would you check against one of these instead, please?

**Comment by Martii** [comment]
> Would you check against one of these instead, please?

Perhaps at a later date... however I seem to recall @sizzlemctwizzle wanted `master` only and that's what I need to go with the documentation. It's a **fixed** restriction.

**Comment by tastytea** [comment]
But why? Isn't the intention to use only the default branch? I think this is a bug that should be fixed.

**Comment by Martii** [comment]
> I think this is a bug that should be fixed.

It's a feature not a bug. Somewhere in the GH issues is a comment on his stance and also mine. Zren, in  a rewrite, made a change to allow importing from any repo which later I found to be a major security hazard and that feature was removed. I would presume more analysis would be needed to see if it could be exploited. Until then it is **fixed**. 

Related issues are #231 and #236 and of course our dependency of `@octokit/rest ᶠᵏᵃ ᵍᶦᵗʰᵘᵇ` which is on hold on our end.

**Comment by monk-time** [comment]
> > I think this is a bug that should be fixed.
> 
> It's a feature not a bug. Somewhere in the GH issues is a comment on his stance and also mine. Zren, in a rewrite, made a change to allow importing from any repo which later I found to be a major security hazard and that feature was removed. I would presume more analysis would be needed to see if it could be exploited. Until then it is **fixed**.
> 
> Related issues are #231 and #236 and of course our dependency of `@octokit/rest ᶠᵏᵃ ᵍᶦᵗʰᵘᵇ` which is on hold on our end.

Found this while trying (unsuccessfully) to understand why #1791 was closed and locked without any explanation (#1781 is extremely confusing and is locked as well). I'm not trying to start any "P.C. debate", only trying to understand the technical considerations at play here. I kindly request that this one won't be locked too, locking without any explanation goes against the clause about "a professional and civil demeanor" in the project's [Terms of Service](https://github.com/OpenUserJS/OpenUserJS.org/blob/master/views/includes/documents/Terms-of-Service.md) that I assume apply to this repo as well.

@Martii, there's nothing in the mentioned issues about default branches. Both me, @tastytea and the reporter in #1781 are trying to sync with the repos that we fully control, not from "any repo" as you mention. The change would be limited to one line in the code that doesn't check anything beyond the name of the updated reference (that includes the branch name). 

Could you please provide any further info on how reducing the restriction to allow `refs/heads/main` (or even `refs/heads/${payload.default_branch}`) can possibly create a security issue? Esp. if `refs/heads/master` doesn't create one. Whereas hardcoding the default branch name denies the service to an increasing amount of repos that have set up a different name, e.g. to any repo created on GitHub after [October 1, 2020](https://github.blog/changelog/2020-10-01-the-default-branch-for-newly-created-repositories-is-now-main/).

I've also failed to find any other comments in the issues that pertain to this topic. I would really appreciate any pointers.
